vm/actor/src/builtin/miner/bitfield_queue.rs

```diff
@@ -0,0 +1,147 @@
+        Ok(Self {
+            amt: Amt::load(root, store)?,
+            quant,
+        })
+    }
+
+    /// Adds values to the queue entry for an epoch.
+    pub fn add_to_queue(
+        &mut self,
+        raw_epoch: ChainEpoch,
+        values: &BitField,
+    ) -> Result<(), Box<dyn StdError>> {
+        if values.is_empty() {
+            // nothing to do.
+            return Ok(());
+        }
+
+        let epoch = self.quant.quantize_up(raw_epoch);
+
+        let bitfield = self
+            .amt
+            .get(epoch as u64)
+            .map_err(|e| format!("failed to lookup queue epoch {}: {:?}", epoch, e))?
+            .ok_or_else(|| actor_error!(ErrNotFound; ""))?;
+
+        self.amt
+            .set(epoch as u64, &bitfield | values)
+            .map_err(|e| format!("failed to set queue epoch {}: {:?}", epoch, e))?;
+
+        Ok(())
+    }
+
+    pub fn add_to_queue_values(
+        &mut self,
+        epoch: ChainEpoch,
+        values: &[u64],
+    ) -> Result<(), Box<dyn StdError>> {
+        self.add_to_queue(epoch, &values.iter().map(|&i| i as usize).collect())
+    }
+
+    /// Cut cuts the elements from the bits in the given bitfield out of the queue,
+    /// shifting other bits down and removing any newly empty entries.
+    ///
+    /// See the docs on `BitField::cut` to better understand what it does.
+    pub fn cut(&mut self, to_cut: &BitField) -> Result<(), String> {
+        let mut epochs_to_remove = Vec::<u64>::new();
+
+        self.amt
+            .for_each_mut(|epoch, bitfield| {
+                *bitfield = bitfield.cut(to_cut);
+
+                if bitfield.is_empty() {
+                    epochs_to_remove.push(epoch);
+                }
+
+                Ok(())
+            })
+            .map_err(|e| format!("failed to cut from bitfield queue: {:?}", e))?;
+
+        // TODO: batch delete
+        for epoch in epochs_to_remove {
+            self.amt.delete(epoch).map_err(|e| {
+                format!("failed to remove empty epochs from bitfield queue: {:?}", e)
+            })?;
+        }
+
+        Ok(())
+    }
+
+    pub fn add_many_to_queue_values(
+        &mut self,
+        values: &HashMap<ChainEpoch, Vec<u64>>,
+    ) -> Result<(), Box<dyn StdError>> {
+        // Update each epoch in-order to be deterministic.
+        // Pre-quantize to reduce the number of updates.
+
+        let mut quantized_values = HashMap::<ChainEpoch, Vec<u64>>::new();
+        let mut updated_epochs = Vec::<ChainEpoch>::new();
+
+        for (&raw_epoch, entries) in values {
+            let epoch = self.quant.quantize_up(raw_epoch);
+            updated_epochs.push(epoch);
+            quantized_values.entry(epoch).or_default().extend(entries);
+        }
+
+        updated_epochs.sort_unstable();
+
+        for epoch in updated_epochs {
+            self.add_to_queue_values(epoch, &quantized_values.remove(&epoch).unwrap_or_default())?;
```

Comment:
Noting that an unwrap to default would just create a discrepancy, but specs-actors would crash so I am fine with this